c7n/resources/account.py

```diff
@@ -424,6 +425,86 @@ def process(self, resources, event=None):
         return []
 
 
+@actions.register('set-password-policy')
+class SetAccountPasswordPolicy(BaseAction):
+    """Set an account's password policy.
+
+    This only changes the policy for the items provided.
+    If this is the first time setting a password policy and an item is not provided it will be
+    set to the defaults defined in the boto docs for IAM.Client.update_account_password_policy
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: set-account-password-policy
+                resource: account
+                filters:
+                  - not:
+                    - type: password-policy
+                      key: MinimumPasswordLength
+                      value: 10
+                      op: ge
+                actions:
+                    - type: set-password-policy
+                      policy:
+                        - key: MinimumPasswordLength
+                          value: 20
+    """
+    schema = type_schema(
+        'set-password-policy',
+        policy={
+            'type': 'array',
+            'items': {
+                'type': 'object',
+                'properties': {
+                    'key': {'type': 'string'},
+                    'value': {'oneOf': [{'type': 'integer'}, {'type': 'boolean'}]}
+                }
+            }
+        })
+    shape = 'PasswordPolicy'
+    permissions = ('iam:GetAccountPasswordPolicy', 'iam:UpdateAccountPasswordPolicy')
+    # We need to whitelist the values accepted by the update call as it is not available via shape
```

Comment:
the shape has all the attributes, not understanding this comment or variable usage

Reply:
The shape does have all of the attributes for PasswordPolicy but from what I can see cannot validate proper params for an operation like update_account_password_policy(). The issue here is that there are distinctions on the get and update operations. For example when getting the password policy, it returns ExpirePasswords when MaxPasswordAge is 0 or not present but the update can only set MaxPasswordAge and will error if ExpirePasswords is provided.


https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html#IAM.Client.get_account_password_policy

Reply:
thats just needing a different shape afaics

Reply:
ie use shape = "UpdateAccountPasswordPolicyRequest"

Reply:
👍 thanks Kapil! I didn't know that was there. Was looking at this document where apparently it's not mentioned. For future reference, how did you find that??

https://github.com/boto/boto3/blob/0cc6042615fd44c6822bd5be5a4019d0901e5dd2/boto3/data/iam/2010-05-08/resources-1.json

Reply:
UTSL ;-) or in this case the json... the structure is generally the same when you get the hang of it, and then instead of seeing 1 and 0s you see...

Reply:
Updated in ba13962

Reply:
I'll still need to strip extraneous fields coming from the getter to make things work. I added a shape_strip function to aws.py to handle this and encourage reuse if the scenario is present in other places.